services/cloudfront/src/main/java/software/amazon/awssdk/services/cloudfront/CloudFrontUtilities.java

```diff
@@ -0,0 +1,753 @@
+     *            <tt>"https://"</tt>
+     * @param privateKey
+     *            The private key data that corresponds to the keypair
+     *            identified by keyPairId
+     * @param keyPairId
+     *            Identifier of a public/private certificate keypair already
+     *            configured in your Amazon Web Services account.
+     * @param expirationDate
+     *            The UTC time and date when the signed URL will expire.
+     * @return A signed URL that will permit access to a specific distribution
+     *         and S3 object.
+     */
+    public static String getSignedUrlWithCannedPolicy(String resourceUrl,
+                                                      PrivateKey privateKey,
+                                                      String keyPairId,
+                                                      Instant expirationDate) {
+        try {
+            String cannedPolicy = buildCannedPolicy(resourceUrl, expirationDate);
+            byte[] signatureBytes = signWithSha1Rsa(cannedPolicy.getBytes(UTF_8), privateKey);
+            String urlSafeSignature = makeBytesUrlSafe(signatureBytes);
+            return resourceUrl
+                   + (resourceUrl.indexOf('?') >= 0 ? "&" : "?")
+                   + "Expires=" + expirationDate.getEpochSecond()
+                   + "&Signature=" + urlSafeSignature
+                   + "&Key-Pair-Id=" + keyPairId;
+        } catch (InvalidKeyException e) {
+            throw SdkClientException.create("Could not sign url", e);
+        }
+    }
+
+    /**
+     * Returns a signed URL that provides tailored access to private content
+     * based on an access time window and an ip range.
+     * For more information, see <a href=
+     * "https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-creating-signed-url-custom-policy.html"
+     * >Creating a signed URL using a custom policy</a>.
+     *
+     * @param protocol
+     *            The protocol of the URL - HTTP or HTTPS
+     * @param distributionDomain
+     *            The domain name of the distribution
+     * @param s3ObjectKey
+     *            The s3 key of the object
+     * @param privateKeyFile
+     *            Your private key file. RSA private key (.pem) and pkcs8 (.der)
+     *            files are supported.
+     * @param keyPairId
+     *            Identifier of a public/private certificate keypair already
+     *            configured in your Amazon Web Services account.
+     * @param activeDate
+     *            The beginning valid date of the signed URL in UTC
+     * @param expirationDate
+     *            The expiration date of the signed URL in UTC
+     * @param ipRange
+     *            The allowed IP address range of the client making the GET
+     *            request, in CIDR form (e.g. 192.168.0.1/24).
+     * @return The signed URL.
+     */
+    public static String getSignedUrlWithCustomPolicy(Protocol protocol,
```

Comment:
I find the combination of input parameters to the two flavors of methods unintuitive (I realized it's probably an inheritance from v1). Do we know why we're combining parts of a url with a private key file as opposed to complete url with a key? Do we need the overload of parts of a url at all?


Do we even need the methods with protocol, distribution domain and object key?

Reply:
Condensed to one variant of each type, which takes a CloudFrontSignerRequest as parameter. Replaced protocol/ distribution domain/object key with resourceUrl, and key file with key